packages/api/src/json-rpc/group/ChainHead.ts

```diff
@@ -0,0 +1,454 @@
+import { BlockHash, Option } from '@dedot/codecs';
+import { Subscription } from '@dedot/providers';
+import {
+  ChainHeadRuntimeVersion,
+  FollowEvent,
+  FollowOperationEvent,
+  MethodResponse,
+  OperationId,
+  RuntimeEvent,
+  StorageQuery,
+  StorageResult,
+} from '@dedot/specs';
+import { Unsub } from '@dedot/types';
+import { assert, Deferred, deferred, ensurePresence, HexString, noop } from '@dedot/utils';
+import { IJsonRpcClient } from '../../types.js';
+import { JsonRpcGroup, JsonRpcGroupOptions } from './JsonRpcGroup.js';
+
+export type OperationHandler<T = any> = {
+  operationId: OperationId;
+  defer: Deferred<T>;
```

Comment:
Keep in mind that you may want to cancel these operations later on.

Reply:
yeah, of course, we do [clean up](https://github.com/dedotdev/dedot/blob/c900236b23e919f26528f22edc5f8c1a1e1883b0/packages/api/src/json-rpc/group/ChainHead.ts#L278-L281) after each operation :)

Reply:
well, no, I meant that you may want to cancel the operation before you receive the response. So, that you can free up operation slots (and resources) for those operations that are no longer needed.

Reply:
That's kind a risky optimization that I think we should be careful of. How do we make sure we will a response if we cancel its operation before we actually receive one? JSON-RPC server implementation might be different between one or the other. What if we interupt an on-going operation and ended up receiving nothing?

Reply:
> That's kind a risky optimization that I think we should be careful of. 

Not it's not. It's a [well documented ](https://paritytech.github.io/json-rpc-interface-spec/api/chainHead_v1_stopOperation.html#chainhead_v1_stopoperation) request that it's meant to free up the resources that are no longer needed.

The point is that sometimes we start a request, and then something happens that makes us no longer interested in that request. The correct thing to do in those cases is to cancel the ongoing operation that is no longer needed, so that we can free up resources and use them for things that are more important.

Keep in mind that for instance, any operations that you started against a block that does no longer belong to the canonical chain are not going to get automatically cancelled, even if you don't need them. So, you probably will want to stop those ones in order to free up resources for the other requests that you have in the queue.

Or for instance, it makes sense when watching changes on a storage key to start races as the blocks come in, and when you receive the response for the most recent block cancel the previous ones.

Or you may need to cancel body searches, etc, etc, etc.

There are plenty of circumstances where you will have to proactively stop an ongoing operation that you are no longer intersted in, so that you can make room for the newer and/or more important ones.